### PR TITLE
Allow test suite extension with keyword `EXTENDS` in settings

### PR DESCRIPTION
**Pull request by okurz**
This allows to inherit existing test suites to prevent repetition.

Related progress issue: https://progress.opensuse.org/issues/55730

### CODE REVIEW

File: lib/OpenQA/WebAPI/Controller/API/V1/Job.pm [hunk 1]
```diff
@@ -817,20 +836,12 @@ sub _generate_job_setting {
...
     # TEST is mandatory, so populate with TestSuite settings.
-    my $test_suites = $schema->resultset('TestSuites')->search(
-        {
-            name => $args->{TEST},
-        });
-
-    if (my $test_suite = $test_suites->next) {
-        my %test_suite_setting = map { $_->key => $_->value } $test_suite->settings;
-
+    if (my %test_suite_setting = %{_test_suite_settings($schema, $args->{TEST})}) {
```
Comment (Martchus) [inline lib/OpenQA/WebAPI/Controller/API/V1/Job.pm]: Why are you creating a copy of the hash here?

File: docs/UsersGuide.asciidoc [hunk 2]
```diff
@@ -159,6 +159,13 @@ and/or os-autoinst's own behavior are:
 * +QEMUVGA+ parameter to declare the video hardware configuration in QEMU
 
 
+==== Test Suite extension
+
+Test suites can extend existing test suites. Specify the test suite you want
+to extend in the variable +EXTENDS+. All variables are inherited from the
```
Comment (Martchus) [inline docs/UsersGuide.asciidoc]: Hopefully nobody is using that variable already for other purposes.

File: lib/OpenQA/WebAPI/Controller/API/V1/Job.pm [hunk 3]
```diff
@@ -759,6 +759,25 @@ sub whoami {
...
+=item _test_suite_settings
+
+Retrieve test suite settings with setting extension from nested test suites
+using the keyword C<EXTENDS>.
+
+=back
+
+=cut
+
+sub _test_suite_settings {
```
Comment (Martchus) [inline lib/OpenQA/WebAPI/Controller/API/V1/Job.pm]: That this appears in the controller for the jobs API routes raises the question: Does this also work when using the isos post route? I suppose it should but I don't see this covered in the tests and there might be (duplicated) code which need to be adapted as well for this to work.

File: docs/UsersGuide.asciidoc (hunk 2 above)
Comment (okurz) [reply docs/UsersGuide.asciidoc]: nope, we checked on osd :)

File: lib/OpenQA/WebAPI/Controller/API/V1/Job.pm (hunk 3 above)
Comment (okurz) [reply lib/OpenQA/WebAPI/Controller/API/V1/Job.pm]: holy moly, there is an ugly duplication in lib/OpenQA/Schema/Result/ScheduledProducts.pm … Let's see if I can fix that, but later

File: lib/OpenQA/WebAPI/Controller/API/V1/Job.pm (hunk 1 above)
Comment (okurz) [reply lib/OpenQA/WebAPI/Controller/API/V1/Job.pm]: you mean I should work on the reference? Ok

**Comment by coolo** [comment]
I'm not sure this makes sense to have now. It's very hidden - and changing test suite A having consequences in test suite B would be super suprising.

We already have enough trouble with changes here affecting other parts in suprising ways - I don't think we need more.

**Comment by okurz** [comment]
I guess this feature does not really add to the confusion where a certain setting came from and how it changed as there are already many places: mediums, machines, worker config, test code settings, test suite and now job templates. Adding a little help for test suites shouldn't be a problem. I guess this might help simplify test suites and identify better candidates for https://progress.opensuse.org/issues/55730

File: lib/OpenQA/WebAPI/Controller/API/V1/Job.pm (hunk 1 above)
Comment (Martchus) [reply lib/OpenQA/WebAPI/Controller/API/V1/Job.pm]: Yes, that's the alternative to the copy.

**Comment by coolo** [comment]
I'm objecting introducing yet another way - the way forward agreed to are job group parameters.

**Comment by Martchus** [comment]
I think if one uses the `EXTENS` feature the expectation actually is that changes from the other testsuite are automatically applied. But we should definitely mention it in the documentation as well. If that is not wanted in certain cases one can still just copy over the parameters.

But the question is indeed whether we actually need this when we have already parameters on job template level and are also able to use references in YAML to avoid duplication.

**Comment by Martchus** [comment]
I think if one uses the `EXTENS` feature the expectation actually is that changes from the other testsuite are automatically applied. But we should definitely mention it in the documentation as well. If that is not wanted in certain cases one can still just copy over the parameters.

But the question is indeed whether we actually need this when we have already parameters on job template level and are also able to use references in YAML to avoid duplication.

**Comment by Martchus** [comment]
I think if one uses the `EXTENS` feature the expectation actually is that changes from the other testsuite are automatically applied. But we should definitely mention it in the documentation as well. If that is not wanted in certain cases one can still just copy over the parameters.

But the question is indeed whether we actually need this when we have already parameters on job template level and are also able to use references in YAML to avoid duplication.

**Comment by Martchus** [comment]
I think if one uses the `EXTENS` feature the expectation actually is that changes from the other testsuite are automatically applied. But we should definitely mention it in the documentation as well. If that is not wanted in certain cases one can still just copy over the parameters.

But the question is indeed whether we actually need this when we have already parameters on job template level and are also able to use references in YAML to avoid duplication.

**Comment by Martchus** [comment]
I think if one uses the `EXTENS` feature the expectation actually is that changes from the other testsuite are automatically applied. But we should definitely mention it in the documentation as well. If that is not wanted in certain cases one can still just copy over the parameters.

But the question is indeed whether we actually need this when we have already parameters on job template level and are also able to use references in YAML to avoid duplication.

**Comment by okurz** [comment]
k

**Pull request closed without merging**